File: summary/elixir/10_cumulative_sum.exs
defmodule CumulativeSum do
  # 累積和を計算する関数
  def cumulative_sum(list) when is_list(list) do
    cumulative_sum(list, 0, [])
  end

  # 再帰的に累積和を計算するヘルパー関数
  defp cumulative_sum([], _, result), do: Enum.reverse(result)
  defp cumulative_sum([head | tail], acc, result) do
    new_acc = acc + head
    cumulative_sum(tail, new_acc, [new_acc | result])
  end

  # 指定された範囲の合計を計算する関数
  def sum_range(list, start, end_) when is_list(list) and is_integer(start) and is_integer(end_) do
    cumulative_sum(list) |> Enum.slice(start - 1..end_ - 1) |> Enum.sum()
  end
end

# テスト
IO.inspect CumulativeSum.cumulative_sum([1, 2, 3, 4, 5])  # [1, 3, 6, 10, 15]
IO.inspect CumulativeSum.cumulative_sum([5, 3, 8, 2, 1])  # [5, 8, 16, 18, 19]

# 1から3までの合計を計算するテスト
IO.inspect CumulativeSum.sum_range([1, 2, 3, 4, 5], 1, 3)
IO.inspect CumulativeSum.sum_range([5, 3, 8, 2, 1], 1, 3)

File: summary/dart/10_cumulative_sum.dart
List<int> cumulativeSum(List<int> data) {
  List<int> accumulates = List.filled(data.length + 1, 0);
  for (int i = 1; i <= data.length; i++) {
    accumulates[i] = accumulates[i - 1] + data[i - 1];
  }
  return accumulates;
}

void printRangeSum(List<int> data, int start, int end) {
  List<int> accumulates = cumulativeSum(data);
  print("Sum from day $start to day $end is ${accumulates[end] - accumulates[start - 1]}");
}

void main() {
  List<int> visitors = [62, 65, 41, 13, 20, 11, 18, 44, 53, 12, 18, 17, 14, 10, 39];
  printRangeSum(visitors, 2, 4);
  printRangeSum(visitors, 4, 13);
}
